### PR TITLE
Added in spatial and median filtering

### PR DESCRIPTION
**Pull request by brechmos-stsci**
I added in spatial filtering which is very similar to the spectral filtering function. This uses the same mechanism of 2D kernels defined in astropy.
I also added in 2D spatial median filtering. 

I ran some basic tests on a dataset I have:
image

CC: @astrofrog

### CODE REVIEW

**Comment by brechmos-stsci** [comment]
I added a second commit that added a function to run a median filter in the spectral direction.
image

**Comment by brechmos-stsci** [comment]
If there is a wish for me to rework the interface for the median filtering, I am happy to do it.

File: spectral_cube/spectral_cube.py [hunk 1]
```diff
@@ -2328,6 +2328,186 @@ def _new_cube_with(self, **kwargs):
...
+    def spatial_smooth_median(self, ksize,
+                        #numcores=None,
+                        **kwargs):
+        """
+        Smooth the cube along the spectral dimension
+
+        Parameters
+        ----------
+        kernel : `~astropy.convolution.Kernel1D`
+            A 1D kernel from astropy
```
Comment (keflavich) [inline spectral_cube/spectral_cube.py]: Spatial smooth should use a 2D kernel, no?

File: spectral_cube/spectral_cube.py [hunk 2]
```diff
@@ -2328,6 +2328,186 @@ def _new_cube_with(self, **kwargs):
...
+        kwargs : dict
+            Passed to the convolve function
+        """
+
+        shape = self.shape
+
+        # "cubelist" is a generator
+        # the boolean check will skip smoothing for bad spectra
+        # TODO: should spatial good/bad be cached?
+        cubelist = ((self.filled_data[ii],
```
Comment (keflavich) [inline spectral_cube/spectral_cube.py]: this is a list of 2D images, no?  so `imagelist` or `planelist` would be better?

File: spectral_cube/spectral_cube.py [hunk 3]
```diff
@@ -2328,6 +2328,186 @@ def _new_cube_with(self, **kwargs):
...
+                     self.mask.include(view=(ii, slice(None), slice(None))))
+                    for ii in range(self.shape[0]))
+
+        pb = ProgressBar(shape[0])
+
+        def _gsmooth_image(args):
+            """
+            Helper function to smooth a spectrum
```
Comment (keflavich) [inline spectral_cube/spectral_cube.py]: an image?

File: spectral_cube/spectral_cube.py [hunk 4]
```diff
@@ -2328,6 +2328,186 @@ def _new_cube_with(self, **kwargs):
...
+            if includemask.any():
+                return ndimage.filters.median_filter(im, size=ksize)
+            else:
+                return im
+
+        # could be numcores, except _gsmooth_spectrum is unpicklable
+        with cube_utils._map_context(1) as map:
+            smoothcube_ = np.array([x for x in
+                                    map(_gsmooth_image, zip(cubelist,
+                                                               itertools.cycle([kwargs]),
```
Comment (keflavich) [inline spectral_cube/spectral_cube.py]: whitespace here; the indentation of `itertools` should match that of `cubelist`

File: spectral_cube/spectral_cube.py [hunk 5]
```diff
@@ -2328,6 +2328,186 @@ def _new_cube_with(self, **kwargs):
...
+                        #numcores=None,
+                        convolve=convolution.convolve,
+                        **kwargs):
+        """
+        Smooth the cube along the spectral dimension
+
+        Parameters
+        ----------
+        kernel : `~astropy.convolution.Kernel1D`
+            A 1D kernel from astropy
```
Comment (keflavich) [inline spectral_cube/spectral_cube.py]: ditto above: 2D

File: spectral_cube/spectral_cube.py (hunk 1 above)
Comment (keflavich) [reply spectral_cube/spectral_cube.py]: In the median case, there is no kernel, so this should be removed.

File: spectral_cube/spectral_cube.py [hunk 6]
```diff
@@ -2328,6 +2328,186 @@ def _new_cube_with(self, **kwargs):
...
+    def spectral_smooth_median(self, ksize,
+                        #numcores=None,
+                        **kwargs):
+        """
+        Smooth the cube along the spectral dimension
+
+        Parameters
+        ----------
+        kernel : `~astropy.convolution.Kernel1D`
+            A 1D kernel from astropy
```
Comment (keflavich) [inline spectral_cube/spectral_cube.py]: no kernel....

File: spectral_cube/spectral_cube.py [hunk 7]
```diff
@@ -2328,6 +2328,186 @@ def _new_cube_with(self, **kwargs):
...
+        convolve : function
+            The astropy convolution function to use, either
+            `astropy.convolution.convolve` or
+            `astropy.convolution.convolve_fft`
```
Comment (keflavich) [inline spectral_cube/spectral_cube.py]: no convolve...

File: spectral_cube/spectral_cube.py (hunk 1 above)
Comment (brechmos-stsci) [reply spectral_cube/spectral_cube.py]: fixed and push coming shortly

File: spectral_cube/spectral_cube.py (hunk 2 above)
Comment (brechmos-stsci) [reply spectral_cube/spectral_cube.py]: fixed and push coming shortly

File: spectral_cube/spectral_cube.py (hunk 3 above)
Comment (brechmos-stsci) [reply spectral_cube/spectral_cube.py]: fixed and push coming shortly

File: spectral_cube/spectral_cube.py (hunk 4 above)
Comment (brechmos-stsci) [reply spectral_cube/spectral_cube.py]: fixed and push coming shortly

File: spectral_cube/spectral_cube.py (hunk 5 above)
Comment (brechmos-stsci) [reply spectral_cube/spectral_cube.py]: fixed and push coming shortly

File: spectral_cube/spectral_cube.py (hunk 6 above)
Comment (brechmos-stsci) [reply spectral_cube/spectral_cube.py]: fixed and push coming shortly

File: spectral_cube/spectral_cube.py (hunk 7 above)
Comment (brechmos-stsci) [reply spectral_cube/spectral_cube.py]: fixed and push coming shortly

**Comment by brechmos-stsci** [comment]
@keflavich I see some fits files referenced in the tests (e.g., adv.fits, vda_beams.fits etc).  Do you have a 3D cube that I should use for the tests, and if so, how could I access it to test if it is not in the `tests/data` directory?

**Comment by keflavich** [comment]
@brechmos-stsci If you run `python setup.py install` (or `python setup.py [anything]`...), it should create cubes in the `tests/data` directory.  If that fails for some reason, raise an Issue, but you can also run `make` in the `tests/data` directory and it will build the test cubes.

**Comment by brechmos-stsci** [comment]
@keflavich I made changes to the code and added several tests to one of the test files.  The tests all appear to hit the `ImportError` due to scipy.ndimage not being available.  Is there something I need to do in the tests in order to flag that it is needed?
```
        if not scipyOK:
>           raise ImportError("Scipy could not be imported: this function won't work.")
E           ImportError: Scipy could not be imported: this function won't work.
```
I am not overly experienced with Travis yet. :-/

**Comment by keflavich** [comment]
Yeah, in the tests you can do:

```python
import pytest
try:
    import scipy.ndimage
    SCIPYOK = True
except ImportError:
    SCIPYOK = False

pytest.skipif('not SCIPYOK')
def test_that_relies_on_scipy():
     pass
```

**Comment by brechmos-stsci** [comment]
@keflavich The two tests that failed, now, are due to import errors trying to do `import signature` which is not related to code I worked on (unless there is some connection I am not aware of).  So, I believe this should be good, but please peruse again. :-)

**Comment by keflavich** [comment]
Oh, I think this issue is that we need to remove py2.7+astropy-dev from the travis build matrix.

**Comment by keflavich** [comment]
@brechmos-stsci I just merged in a test grid update in #431.  Could you update against that (either merge or rebase)?  That should fix the tests

**Comment by brechmos-stsci** [comment]
I merged your test grid update changes and the Travis tests ran. If there are any other changes / updates you would like, please tag me. Happy to make them.

**Pull request merged**